crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -475,6 +563,123 @@ impl WritebackCache {
+        for (i, object) in objects.into_iter().enumerate() {
+            if let Some(object) = object {
+                result.push(object);
+            } else {
+                return Err(SuiError::UserInputError {
+                    error: UserInputError::ObjectNotFound {
+                        object_id: object_ids[i],
+                        version: None,
+                    },
+                });
+            }
+        }
+        Ok(result)
+    }
+
+    fn verify_live_object(obj_ref: &ObjectRef, live_object: &Object) -> SuiResult {
+        if obj_ref.1 != live_object.version() {
+            debug!(
+                "object version unavailable for consumption: {:?} (current: {})",
+                obj_ref,
+                live_object.version()
+            );
+            return Err(SuiError::UserInputError {
+                error: UserInputError::ObjectVersionUnavailableForConsumption {
+                    provided_obj_ref: *obj_ref,
+                    current_version: live_object.version(),
+                },
+            });
+        }
+
+        let live_digest = live_object.digest();
+        if obj_ref.2 != live_digest {
+            debug!("object digest mismatch: {:?} vs {:?}", obj_ref, live_digest);
+            return Err(SuiError::UserInputError {
+                error: UserInputError::InvalidObjectDigest {
+                    object_id: obj_ref.0,
+                    expected_digest: live_digest,
+                },
+            });
+        }
+
+        Ok(())
+    }
+
+    fn clear_cached_locks(&self, locks: &[(ObjectRef, LockDetails)]) {
+        for (obj_ref, lock) in locks {
+            let entry = self.locked_transactions.entry(*obj_ref);
+            let occupied = match entry {
+                DashMapEntry::Vacant(_) => panic!("lock must exist"),
+                DashMapEntry::Occupied(occupied) => occupied,
+            };
+
+            if occupied.get() == lock {
+                trace!("clearing lock: {:?}", lock);
+                occupied.remove();
+            } else {
+                // this is impossible because the only case in which we overwrite a
+                // lock is when the lock is from a previous epoch. but we are holding
+                // execution_lock, so the epoch cannot have changed.
+                panic!("lock was changed since we set it");
+            }
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn acquire_transaction_locks(
+        &self,
+        epoch_store: &AuthorityPerEpochStore,
+        owned_input_objects: &[ObjectRef],
+        transaction: VerifiedSignedTransaction,
+    ) -> SuiResult {
+        let tx_digest = *transaction.digest();
+
+        let mut locks_to_write: Vec<(_, LockDetails)> =
+            Vec::with_capacity(owned_input_objects.len());
+
+        let object_ids = owned_input_objects.iter().map(|o| o.0).collect::<Vec<_>>();
+        let live_objects = self.multi_get_objects_must_exist(&object_ids)?;
```

Comment:
What is the reason to use multi_get_objects instead of reading the `live_owned_object_markers` table, which we did in the passthrough cache impl?

Reply:
because we want to try the cache first instead of going straight to the db?

i could add a new method that reads the live objects from the cache, and falls back to the live object marker to avoid doing any db scans. Do you think thats worth it right now? 

(There's no point in mirroring the live object marker table in the cache, because getting the latest live object from the cache is very cheap.)